array.c

```diff
@@ -6146,7 +6144,7 @@ flatten(VALUE ary, int level)
 	st_clear(memo);
     }
 
-    RBASIC_SET_CLASS(result, rb_obj_class(ary));
+    RBASIC_SET_CLASS(result, klass);
```

Comment:
Is this still necessary? From `flatten` this is `rb_cArray` (and that was set by `ary_new`) and from `flatten!` the result will be the argument for `replace` so the exact subclass of `Array` shouldn't matter, right?

Reply:
It's necessary to set the class.  The `ary_new` call creates a hidden array.  However, I'll refactor the function to not take a `klass` and instead always use `rb_cArray`.

Reply:
Are you sure? I see that `rb_ary_s_create` and `rb_ary_times`, for example, use `ary_new` and return that directly...

Reply:
`ary_new`'s first argument is the class, and if you pass 0, it creates a hidden array.  Both `rb_ary_s_create` and `rb_ary_times` pass in a class object and not 0.

Reply:
Ah, got it, thanks 👍